### PR TITLE
introduces MachineDeletionHooksController

### PR DESCRIPTION
**Pull request by p0lyn0mial**
The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allows this operator to hold up the removal of a machine

### CODE REVIEW

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController reconciles machine hooks for master machines
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
```
Comment (p0lyn0mial) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:41]: I am going to wire this controller in https://github.com/openshift/cluster-etcd-operator/pull/779

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		memberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		currentMemberIPListSet.Insert(memberIP)
+	}
+
+	// find a machine that doesn't have a member
+	var errs []error
+	for machinePendingDeletionIP, machinePendingDeletion := range ceohelpers.IndexMachinesByNodeInternalIP(machinesPendingDeletion) {
```
Comment (p0lyn0mial) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:120]: `ceohelpers.IndexMachinesByNodeInternalIP` will be available once https://github.com/openshift/cluster-etcd-operator/pull/780 merges

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go (hunk 1 above)
Comment (deads2k) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:41]: > I am going to wire this controller in #779

Without wiring, we cannot verify that this code works.

**Comment by deads2k** [comment]
/approve

This needs to be wired in order to have CI test it.

**Comment by p0lyn0mial** [comment]
rebased against the `master`, just a small change to https://github.com/openshift/cluster-etcd-operator/pull/781/commits/a7a913a1c677968537666612392570b7d76d0660 (removed `machineAPI` as it is already defined)

**Pull request merged**